File: linked_list/LL_heap_pq/minheap_pair.c
#include <stdio.h>
#include <stdlib.h>

#define HEAP_SIZE 100

// Pair 구조체: 두 개의 정수 x, y를 저장
typedef struct {
    int x, y;
} Pair;

// MinHeap 구조체: Pair를 저장하는 최소 힙
typedef struct {
    Pair elements[HEAP_SIZE];
    int size; // 현재 힙의 크기
} MinHeap;

// 함수 프로토타입
MinHeap createMinHeap();
void minHeapInsert(MinHeap* heap, Pair value);
Pair minHeapPop(MinHeap* heap);
void printMinHeap(MinHeap* heap);
void bubbleUpMin(MinHeap* heap, int index);
void bubbleDownMin(MinHeap* heap, int index);
void minHeapify(MinHeap* heap, int size, int index);
void sortMinHeap(MinHeap* heap);
int comparePairsMin(Pair a, Pair b);

// 빈 최소 힙 생성
MinHeap createMinHeap() {
    MinHeap heap = {.size = 0};
    return heap;
}

// 최소 힙에 요소 삽입
void minHeapInsert(MinHeap* heap, Pair value) {
    if (heap->size >= HEAP_SIZE) {
        printf("Min-Heap is full!\n");
        return;
    }
    heap->elements[heap->size] = value; // 배열의 마지막에 삽입
    bubbleUpMin(heap, heap->size);      // 힙 속성 복구
    heap->size++;
}

// 최소값을 제거하고 반환
Pair minHeapPop(MinHeap* heap) {
    if (heap->size == 0) {
        printf("Min-Heap is empty!\n");
        return (Pair){-1, -1};
    }
    Pair minValue = heap->elements[0];   // 루트 요소(최소값)
    heap->elements[0] = heap->elements[--heap->size]; // 마지막 요소를 루트로 이동
    bubbleDownMin(heap, 0);             // 힙 속성 복구
    return minValue;
}

// 힙의 모든 요소를 출력
void printMinHeap(MinHeap* heap) {
    for (int i = 0; i < heap->size; i++) {
        printf("(%d, %d) ", heap->elements[i].x, heap->elements[i].y);
    }
    printf("\n");
}

// 두 Pair를 비교하여 우선순위를 결정 (x 우선, 같으면 y 비교)
int comparePairsMin(Pair a, Pair b) {
    if (a.x != b.x) {
        return a.x - b.x; // x가 작은 Pair 우선
    }
    return a.y - b.y; // x가 같으면 y가 작은 Pair 우선
}

// 삽입 시 최소 힙 속성을 유지하도록 상향 이동
void bubbleUpMin(MinHeap* heap, int index) {
    while (index > 0 && comparePairsMin(heap->elements[index], heap->elements[(index - 1) / 2]) < 0) {
        Pair temp = heap->elements[index];
        heap->elements[index] = heap->elements[(index - 1) / 2];
        heap->elements[(index - 1) / 2] = temp;
        index = (index - 1) / 2;
    }
}

// 삭제 시 최소 힙 속성을 유지하도록 하향 이동
void bubbleDownMin(MinHeap* heap, int index) {
    int size = heap->size;
    while (index * 2 + 1 < size) {
        int child = index * 2 + 1;
        if (child + 1 < size && comparePairsMin(heap->elements[child + 1], heap->elements[child]) < 0) {
            child++;
        }
        if (comparePairsMin(heap->elements[index], heap->elements[child]) <= 0) {
            break;
        }
        Pair temp = heap->elements[index];
        heap->elements[index] = heap->elements[child];
        heap->elements[child] = temp;
        index = child;
    }
}

// 주어진 노드에서 시작하여 최소 힙 속성을 복구
void minHeapify(MinHeap* heap, int size, int index) {
    int smallest = index;
    int left = 2 * index + 1;
    int right = 2 * index + 2;

    if (left < size && comparePairsMin(heap->elements[left], heap->elements[smallest]) < 0) {
        smallest = left;
    }
    if (right < size && comparePairsMin(heap->elements[right], heap->elements[smallest]) < 0) {
        smallest = right;
    }
    if (smallest != index) {
        Pair temp = heap->elements[index];
        heap->elements[index] = heap->elements[smallest];
        heap->elements[smallest] = temp;
        minHeapify(heap, size, smallest);
    }
}

// 힙 요소를 오름차순으로 정렬 (힙 정렬 사용)
void sortMinHeap(MinHeap* heap) {
    int originalSize = heap->size;

    // 최소 힙 구조로 변환
    for (int i = heap->size / 2 - 1; i >= 0; i--) {
        minHeapify(heap, heap->size, i);
    }

    // 힙에서 가장 작은 요소를 배열 뒤로 이동하여 정렬
    for (int i = heap->size - 1; i > 0; i--) {
        Pair temp = heap->elements[0];
        heap->elements[0] = heap->elements[i];
        heap->elements[i] = temp;

        minHeapify(heap, i, 0);
    }

    // 원래 크기 복구
    heap->size = originalSize;
}

// 메인 함수: 구현 테스트
int main() {
    printf("Min-Heap with Pairs Implementation:\n");

    MinHeap heap = createMinHeap();

    minHeapInsert(&heap, (Pair){10, 20});
    minHeapInsert(&heap, (Pair){15, 25});
    minHeapInsert(&heap, (Pair){5, 10});
    minHeapInsert(&heap, (Pair){10, 15});

    printf("Min-Heap after insertions: ");
    printMinHeap(&heap);

    printf("\nSorting the heap:\n");
    sortMinHeap(&heap);
    printMinHeap(&heap);

    return 0;
}

File: linked_list/LL_heap_pq/minheap_pair.py
class MinHeap:
    def __init__(self):
        """
        MinHeap 초기화.
        내부적으로 요소를 저장할 빈 리스트 생성.
        """
        self.elements = []

    def compare(self, pair1, pair2):
        """
        두 Pair를 비교하여 우선순위를 결정.
        - x 값이 작은 Pair가 우선.
        - x 값이 같으면 y 값이 작은 Pair가 우선.
        """
        if pair1[0] != pair2[0]:
            return pair1[0] < pair2[0]
        return pair1[1] < pair2[1]

    def bubble_up(self, index):
        """
        새로운 요소 삽입 후 상향 이동하며 Min-Heap 속성 유지.
        부모 노드와 비교해 조건에 맞으면 교환.
        """
        while index > 0 and self.compare(self.elements[index], self.elements[(index - 1) // 2]):
            self.elements[index], self.elements[(index - 1) // 2] = self.elements[(index - 1) // 2], self.elements[index]
            index = (index - 1) // 2

    def bubble_down(self, index):
        """
        루트에서 요소를 제거한 후 하향 이동하며 Min-Heap 속성 유지.
        자식 노드와 비교해 조건에 맞으면 교환.
        """
        size = len(self.elements)
        while index * 2 + 1 < size:
            child = index * 2 + 1
            if child + 1 < size and self.compare(self.elements[child + 1], self.elements[child]):
                child += 1
            if self.compare(self.elements[index], self.elements[child]):
                break
            self.elements[index], self.elements[child] = self.elements[child], self.elements[index]
            index = child

    def min_heapify(self, size, index):
        """
        특정 노드에서 시작하여 서브트리를 Min-Heap 구조로 복구.
        - `sort` 메서드에서 사용.
        """
        smallest = index
        left = 2 * index + 1
        right = 2 * index + 2

        if left < size and self.compare(self.elements[left], self.elements[smallest]):
            smallest = left
        if right < size and self.compare(self.elements[right], self.elements[smallest]):
            smallest = right

        if smallest != index:
            self.elements[index], self.elements[smallest] = self.elements[smallest], self.elements[index]
            self.min_heapify(size, smallest)

    def insert(self, pair):
        """
        새로운 Pair 삽입.
        리스트 끝에 추가한 후 `bubble_up` 호출로 Min-Heap 속성 유지.
        """
        self.elements.append(pair)
        self.bubble_up(len(self.elements) - 1)

    def pop(self):
        """
        Min-Heap의 최소값(Pair)을 제거하고 반환.
        루트 요소 제거 후 `bubble_down` 호출로 Min-Heap 속성 유지.
        """
        if not self.elements:
            raise IndexError("Min-Heap is empty!")
        min_pair = self.elements[0]
        self.elements[0] = self.elements.pop()
        self.bubble_down(0)
        return min_pair

    def sort(self):
        """
        Min-Heap의 요소를 내림차순으로 정렬.
        - `min_heapify`를 사용하여 힙 정렬 방식으로 정렬.
        """
        original_size = len(self.elements)
        # Min-Heap 구조로 변환
        for i in range(original_size // 2 - 1, -1, -1):
            self.min_heapify(original_size, i)

        # 힙 정렬 수행
        for i in range(original_size - 1, 0, -1):
            self.elements[0], self.elements[i] = self.elements[i], self.elements[0]
            self.min_heapify(i, 0)

    def print(self):
        """
        현재 Min-Heap의 모든 요소를 출력.
        """
        print(self.elements)


# Min-Heap 테스트
print("Min-Heap with Pairs Implementation:")
min_heap = MinHeap()
min_heap.insert((10, 20))
min_heap.insert((15, 25))
min_heap.insert((5, 10))
min_heap.insert((10, 15))

print("Min-Heap after insertions: ", end="")
min_heap.print()

print("\nSorting the heap:")
min_heap.sort()
min_heap.print()

print(f"\nPopping minimum: {min_heap.pop()}")
print("Min-Heap after popping: ", end="")
min_heap.print()